lib/main.dart

```diff
@@ -0,0 +1,189 @@
+          toolCache.cacheDir(extractedFolder, toolName, version, architecture));
+    }
+
+    final pubCache = path.join(
+        getProcessEnv(os == 'windows' ? 'USERPROFILE' : 'HOME')!, '.pub-cache');
+
+    core.exportVariable('DART_HOME', sdkPath);
+    core.addPath(path.join(sdkPath, 'bin'));
+    core.exportVariable('PUB_CACHE', pubCache);
+    core.addPath(path.join(pubCache, 'bin'));
+
+    // Create the OIDC token used for pub.dev publishing.
+    await createPubOIDCToken();
+
+    // Configure the outputs.
+    if (raw) {
+      core.setOutput('dart-version', getVersionFromSdk(sdkPath)!);
+    } else {
+      core.setOutput('dart-version', version);
+    }
+
+    // Report success; print version.
+    await promiseToFuture(exec.exec('dart', ['--version']));
+  } catch (error) {
+    core.setFailed('$error');
+  }
+}
+
+String? getVersionFromSdk(String sdkPath) {
+  final versionFilePath = path.join(sdkPath, 'version');
+  print(versionFilePath);
+
+  if (fs.existsSync(versionFilePath)) {
+    return fs.readFileSync(versionFilePath, 'utf8').trim();
+  } else {
+    return null;
+  }
+}
+
+/// Returns 'x64', 'ia32', 'arm', or 'arm64'.
+String getArch() {
+  const supported = ['x64', 'ia32', 'arm', 'arm64'];
+  return supported.contains(os.arch()) ? os.arch() : 'x64';
+}
+
+/// Returns 'linux', 'windows', or 'macos'.
+String getPlatform() {
+  if (os.platform() == 'win32') return 'windows';
+  return os.platform() == 'darwin' ? 'macos' : 'linux';
+}
+
+// When enabled through env variables, create OIDC token for publishing on
+// pub.dev.
+Future<void> createPubOIDCToken() async {
+  final tokenRequestUrl = getProcessEnv('ACTIONS_ID_TOKEN_REQUEST_URL');
+  final tokenRequestToken = getProcessEnv('ACTIONS_ID_TOKEN_REQUEST_TOKEN');
+
+  if (tokenRequestUrl == null || tokenRequestToken == null) {
+    return;
+  }
+
+  final token = await promiseToFuture(core.getIDToken('https://pub.dev'));
+
+  core.exportVariable('PUB_TOKEN', token);
+
+  await promiseToFuture(exec.exec('dart',
+      ['pub', 'token', 'add', 'https://pub.dev', '--env-var', 'PUB_TOKEN']));
+}
+
+// https://storage.googleapis.com/dart-archive/channels/stable/release/latest/VERSION
+// {
+//   "date": "2023-02-07",
+//   "version": "2.19.2",
+//   "revision": "e46b4f59490230778e907bde2eedb06b062d31be"
+// }
+
+// Query google storage for the most recent published SDK version for the given
+// channel and flavor.
+Future<String?> latestPublishedVersion(String channel, String flavor) async {
+  final url = 'https://storage.googleapis.com/dart-archive/channels/'
+      '$channel/$flavor/latest/VERSION';
+
+  final http = HttpClient('setup-dart', [], {
+    'allowRedirects': true,
+    'maxRedirects': 3,
+    'allowRetries': true,
+    'maxRetries': 3,
+  });
+
+  JSObject response = await promiseToFuture(http.getJson(url));
+  JSObject? result = getProperty(response, 'result');
+  return result == null ? null : getProperty(result, 'version');
```

Comment:
Given the call sites assert non-nullability, we could just make this non-nullable and move the assert here.

Reply:
+1 - I changed this to return a non-null value, and instead throw an exception if we can't retrieve the latest version